### ISSUE TITLE
why this point not be played?

### ISSUE THREAD

**Issue by ericyan71** [leela-zero/leela-zero#2445]
hello, when I use lizzie to play a game, I found a strange thing, black should kill a white dragon, but the point which will kill dragon has never to be simulated.

weight:  232 - 06e03050
gpu: 1080ti with cuda 10.0, cudn 7.4

![20190713132531](https://user-images.githubusercontent.com/22901866/61167481-4e4c7880-a572-11e9-84f6-15b527e7fbf2.jpg)

**Comment by nemja** [comment]
It likely thinks that move is illegal. An old bug #877.

**Comment by ericyan71** [comment]
i checked heatmap, it's value was zero indeed, has it been fixed?
![2](https://user-images.githubusercontent.com/22901866/61179114-86b18c80-a62e-11e9-9659-c2ddb4c986d9.png)

**Comment by ihavnoid** [comment]
Hi, do we have a sgf for this game?

**Comment by nemja** [comment]
This was not fixed AFAIK, the root of the problem is a misunderstanding about how AG worked (loss function on illegal moves).

But in this particular case any zero bot may err even with correct training code. Note how long and curly this white string is, therefore its liberty count is not reliable. And should it have +1 lib somewhere the move would really be illegal.

**Comment by ericyan71** [comment]
is it not a problem? why needn't fix it，I think LZ wil much morel stronger after fix this problem thanks.

**Comment by roy7** [comment]
Someone on discord just saw the same thing happen. I asked them to post it here.

**Comment by roy7** [comment]
Someone would need to change heat map to print lots of decimal places but are moves truly zero? Does that mean since it is multiplicative it is impossible for the move to be explored?

https://cdn.discordapp.com/attachments/417022162348802050/627656292323819558/image0.png

**Comment by ihavnoid** [comment]
I think one possible fix is to put a small amount of delta so that even if the policy is zero we still retain a very small smount of probability.

Another rather hacky fix is to put a certain amount of probability if the position is seemingly illegal, but will turn legal simply because we capture some opponent stones.

**Comment by tapsika** [comment]
The real problem is not the noticeably erratic behaviour in rare cases like this, but the half stone (?) or so strength that may be lost by overloading nets with useless task of predicting legality (and even depending on that guess being correct in one direction). Even without hints in DM papers I would doubt they did this as they went for max strength.

Since fixing this would need retraining, imo the correct thing atm would be to fix the training code with a config switch that can be set to old behavior as default. That would at least allow @bubblesld or someone else who trains from scratch to try it both ways, and compare the strength lost (#2504 #2192).

Though even that would be problematic since the selfplay / training data itself may be a bit polluted by this at the moment. But while there were reports that fixing this is a huge help in other projects, it is unclear how much impact this have on the max strength achievable with a given size, so getting any data on this could be useful.

**Comment by tapsika** [comment]
I took a quick look at LZ training code, and the problem seems to be deeper rooted than illegal moves. Apparently LZ pulls ANY move untried during search towards 0 policy. Which seems wrong since - as these moves were not tried - we gained zero information about them, yet they are treated exactly as if they would have been tried again and found very bad.

I think there should at least be an adjustable training weight factor for such moves (with weight almost certainly lower than 1).

**Comment by ihavnoid** [comment]
@tapsika - when we do the self-plays we add noise to the policy values precisely for this reason.  It's just that the policy is fixed to zero when we actually play a game.

See UCTNode::prepare_root_node

**Comment by ihavnoid** [comment]
Anyway, I am trying to create a fix to address this specific weak point.  I am not sure if this worthy merging here, but at least we can figure out if this is the problem or not.

Can anybody provide a sgf file that shows this issue?

**Comment by tapsika** [comment]
I don't think noise forces all legal moves to be explored (this won't fit into the low selfplay visit limit anyway). And as long as there are unexplored moves, pulling them to 0 without any new info seems wrong, at least with same weight (blindly reinforces earlier beliefs, also training without new data = overfitting earlier data),

**Comment by ihavnoid** [comment]
Pulling everything up by a small arbitrary amount doesn't look right either - even putting a very small 0.001 on every unexplored move already adds so much misinformation that we have to search every single possible move and turns the whole policy output useless.

Anyway, usually in these cases somebody has to prove themselves (with hard data) that their idea is right, otherwise nothing changes.

**Comment by tapsika** [comment]
I feel we are talking about different things. I meant the training targets from (current, unchanged) selfplay search, not some kind of fix for the wrong policy later. For untried moves I see two training options: pull them towards 0 policy, or leave them unchanged (no gradient for them since we got no actual training data for them). An adjustable training weight (if moves with real search result weight=1) seems a natural solution.

Proving with hard data is not easy though since this needs retraining from scratch.

Btw, for later search a policy minimum or some other way to slowly start exploring all legal moves seems logical, in fact I think there were other attempts to reduce the reliance on policy as visits accumulate. Such will naturally fix the noticeable blind spots, at least in high visit games. But this is not the same as fixing the training (network strength can suffer from the incorrect training even if the resulting blind spot will be covered up later).

**Comment by LL145** [comment]
SGF for this Two-Headed Dragon game

(  ;GM[1]FF[4]SZ[19]GN[第28期日本龙星战决赛<绝艺讲解>]DT[2019-09-23]PB[上野爱咲美]PW[一力辽]BR[P3段]WR[P8段]KM[650]HA[0]RU[Japanese]AP[Sabaki:0.43.1]RE[W+R]TM[600]TC[1]TT[30]RL[0]CA[UTF-8]  ;B[pd]  ;W[dp]  ;B[qp]  ;W[dd]  ;B[op]  ;W[qf]  ;B[qh]  ;W[md]  ;B[ne]  ;W[qc]  ;B[qd]  ;W[pc]  ;B[nd]  ;W[nc]  ;B[rc]  ;W[oc]  ;B[re]  ;W[me]  ;B[mf]  ;W[nf]  ;B[of]  ;W[ng]  ;B[og]  ;W[lf]  ;B[mg]  ;W[nh]  ;B[lg]  ;W[oh]  ;B[ph]  ;W[oj]  ;B[kf]  ;W[kd]  ;B[qj]  ;W[li]  ;B[ji]  ;W[kh]  ;B[jd]  ;W[jc]  ;B[ke]  ;W[ld]  ;B[kg]  ;W[jh]  ;B[id]  ;W[ic]  ;B[if]  ;W[ii]  ;B[fq]  ;W[hd]  ;B[he]  ;W[jg]  ;B[jf]  ;W[ie]  ;B[je]  ;W[gd]  ;B[hh]  ;W[gf]  ;B[hf]  ;W[jj]  ;B[dn]  ;W[en]  ;B[em]  ;W[fn]  ;B[cq]  ;W[dq]  ;B[cp]  ;W[do]  ;B[co]  ;W[cr]  ;B[dm]  ;W[er]  ;B[jq]  ;W[bm]  ;B[bn]  ;W[cm]  ;B[cn]  ;W[cj]  ;B[gp]  ;W[go]  ;B[gl]  ;W[hp]  ;B[hn]  ;W[gq]  ;B[jo]  ;W[gj]  ;B[ej]  ;W[fh]  ;B[gi]  ;W[fi]  ;B[fj]  ;W[hj]  ;B[ge]  ;W[fe]  ;B[gg]  ;W[ff]  ;B[le]  ;W[dh]  ;B[nj]  ;W[nk]  ;B[mj]  ;W[od]  ;B[oe]  ;W[mi]  ;B[mk]  ;W[oi]  ;B[ok]  ;W[pk]  ;B[nl]  ;W[pj]  ;B[ik]  ;W[rh]  ;B[ri]  ;W[qk]  ;B[qi]  ;W[hl]  ;B[il]  ;W[fl]  ;B[gk]  ;W[hk]  ;B[gm]  ;W[fm]  ;B[fk]  ;W[rl]  ;B[qm]  ;W[rj]  ;B[sj]  ;W[sk]  ;B[rk]  ;W[im]  ;B[kk]  ;W[rj]  ;B[si]  ;W[ih]  ;B[gh]  ;W[ij]  ;B[kj]  ;W[ki]  ;B[rm]  ;W[jk]  ;B[jl]  ;W[hm]  ;B[gn]  ;W[kl]  ;B[jm]  ;W[jn]  ;B[km]  ;W[kn]  ;B[lm]  ;W[in]  ;B[ho]  ;W[fp]  ;B[io]  ;W[pg]  ;B[pe]  ;W[qg]  ;B[rg]  ;W[ln]  ;B[lp]  ;W[pm]  ;B[mm]  ;W[mp]  ;B[mq]  ;W[rf]  ;B[sh]  ;W[sf]  ;B[rk]  ;W[lj]  ;B[lk]  ;W[rj]  ;B[se]  ;W[qo]  ;B[ro]  ;W[pn]  ;B[po]  ;W[sg]  ;B[rk]  ;W[lq]  ;B[lo]  ;W[rj]  ;B[qe]  ;W[mn]  ;B[no]  ;W[mo]  ;B[nq]  ;W[kq]  ;B[jp]  ;W[qn]  ;B[rn]  ;W[nn]  ;B[oo]  ;W[ol]  ;B[nm]  ;W[on]  ;B[ll]  ;W[ql]  ;B[cd]  ;W[cc]  ;B[dc]  ;W[bc]  ;B[kc]  ;W[kb]  ;B[bd]  ;W[ec]  ;B[db]  ;W[eb]  ;B[ch]  ;W[di]  ;B[de]  ;W[ed]  ;B[dg]  ;W[cg]  ;B[bb]  ;W[cb]  ;B[ac]  ;W[da]  ;B[bg]  ;W[cf]  ;B[bi]  ;W[bj]  ;B[om]  ;W[nk]  ;B[bf]  ;W[dk]  ;B[df]  ;W[ci]  ;B[bh]  ;W[el]  ;B[ok]  ;W[rh])

**Comment by ihavnoid** [comment]
@LL145 - what is the intention?  What happens if you evaluate this sgf but what do you expect?

**Comment by LL145** [comment]
@ihavnoid 
this is the sgf for this position (This situation is not an actual game):
https://cdn.discordapp.com/attachments/417022162348802050/627656292323819558/image0.png
I find this blind spot when analyzing this game

**Comment by roy7** [comment]
@LL145 Do you recall which network you were using at the time?

**Comment by LL145** [comment]
@roy7 
LZ245
